### ISSUE TITLE
Support for publicNetworkAccess in storage_account

### ISSUE THREAD

**Issue by jrhunger** [hashicorp/terraform-provider-azurerm#16335]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The underlying Storage Account object has an optional publicNetworkAccess parameter, with possible values Enabled or Disabled.
This is not currently surfaced to azurerm_storage_account.

In an enterprise setting, if a policy is applied requiring publicNetworkAccess= Disabled for all storage accounts, then there is no good way to create storage accounts using Terraform.  (There is a bad way to do it - by including ARM templates)

### New or Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Potential Terraform Configuration

```hcl
resource "azurerm_storage_account" "example" {
  name                          = "storageaccountname"
  resource_group_name           = azurerm_resource_group.example.name
  location                      = azurerm_resource_group.example.location
  account_tier                  = "Standard"
  account_replication_type      = "GRS"
  public_network_access_enabled = false

  tags = {
    environment = "staging"
  }
}
```

### References

https://docs.microsoft.com/en-us/azure/templates/microsoft.storage/storageaccounts?tabs=json#storageaccountpropertiescreateparameters
https://docs.microsoft.com/en-us/rest/api/storagerp/storage-accounts/create#publicnetworkaccess

**Comment by tombuildsstuff** [comment]
hi @jrhunger 

Thanks for opening this issue.

Support for this is available [via the `allow_nested_items_to_be_public` field in the `azurerm_storage_account` resource](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_account#allow_nested_items_to_be_public) - as such support for this is already available.

Thanks!

**Comment by kmehkeri** [comment]
These are two different options available via Azure CLI:

* `--allow-blob-public-access` - this is what Terraform sets via its `allow_nested_items_to_be_public` option
* `--public-network-access` - this is what @jrhunger wants (and me too)

**Comment by jrhunger** [comment]
@tombuildsstuff is it possible to re-open this issue, or should i open a new one? 
Elaborating on the previous comment, there are two different uses of the word "public"

* allowBlobPublicAccess (allow_nested_items_to_be_public) - public here means anonymous, without authentication to the API
* publicNetworkAccess - public here means public networks aka the internet. disabling means that this storage account can only be accessed through private endpoints

excerpt from [this page](https://docs.microsoft.com/en-us/azure/storage/common/storage-network-security?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&tabs=azure-cli#change-the-default-network-access-rule):

> To block traffic from all networks, use the az storage account update command and set the --public-network-access parameter to Disabled. Traffic will be allowed only through a [private endpoint](https://docs.microsoft.com/en-us/azure/storage/common/storage-private-endpoints). You'll have to create that private endpoint.

**Comment by tombuildsstuff** [comment]
@jrhunger @kmehkeri thanks for the clarification, yeah this makes sense to reopen 👍

**Comment by Chambras** [comment]
The workaround I found is to use the new [AzAPI provider](https://docs.microsoft.com/en-us/azure/developer/terraform/overview-azapi-provider)

It would be something like this

```terraform
resource "azapi_update_resource" "example" {
  type        = "Microsoft.Storage/storageAccounts@2021-09-01"
  resource_id = azurerm_storage_account.example.id

  body = jsonencode({
    properties = {
      publicNetworkAccess = "Disabled"
    }
  })
}

```
Would be nice to have native support though :) 
Hope this helps

cc @kmehkeri @jrhunger @tombuildsstuff

**Comment by kmehkeri** [comment]
I tried to add the support. Throwing in an extra parameter is quite straightforward, however it seems that Azure Storage API version upgrade would be required - current used by azurerm is [2021-04-01](https://docs.microsoft.com/en-us/azure/templates/microsoft.storage/2021-04-01/storageaccounts?tabs=bicep), while `publicNetworkAccess` attribute is supported from [2021-06-01](https://docs.microsoft.com/en-us/azure/templates/microsoft.storage/2021-06-01/storageaccounts?tabs=bicep). Latest version available as of now is [2021-09-01](https://docs.microsoft.com/en-us/azure/templates/microsoft.storage/2021-09-01/storageaccounts?tabs=bicep).

I don't feel confident doing this upgrade myself. I found an earlier upgrade [here](https://github.com/hashicorp/terraform-provider-azurerm/pull/14083), which could be used as a guide, but honestly it's hard to extract diff from Azure docs about what exactly changed between versions, so I'm not sure what would break.

If it's reasonable I can raise a WIP PR with what I have already.

**Comment by ltalirz** [comment]
> The workaround I found is to use the new [AzAPI provider](https://docs.microsoft.com/en-us/azure/developer/terraform/overview-azapi-provider)
> 
> It would be something like this
> 
> ```terraform
> resource "azapi_update_resource" "example" {
>   type        = "Microsoft.Storage/storageAccounts@2021-09-01"
>   resource_id = azurerm_storage_account.example.id
> 
>   body = jsonencode({
>     properties = {
>       publicNetworkAccess = "Disabled"
>     }
>   })
> }
> ```

Sorry for the newbie question, but is this workaround indeed supposed to enable the creation of the storage account via terraform, or is it just an illustration of how to update the storage account _after_ it has already been created?

I have added this workaround to my `.tf` script (and updated the variables accordingly), and somehow I am still stopped by the rule requiring public access to be disallowed.

**Comment by skri547** [comment]
@kmehkeri  @tombuildsstuff  Any idea on when this functionality will be available? we are also waiting on this from few days.

**Comment by michaelrechani** [comment]
Any update on this?  Definitely important to have as this is a security configuration.

**Comment by kmehkeri** [comment]
It looks like this has been now added by someone else in #18005, it is available since v3.21.0.

**Comment by neilmca-inc** [comment]
I tried this in 3.27.0 and it worked successfully

**Comment by Chambras** [comment]
Cool! It seems the documentation has been updated as well:
https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_account#public_network_access_enabled

I tried with terraform 1.3.2 and azurerm 3.27.0 and worked!!!
Awesome great job! 
I guess we can close this one now :) 
cc. @tombuildsstuff

**Comment by jrhunger** [comment]
Working for us as well.
